packages/contracts/contracts/optimistic-ethereum/ovm/StateManagerGasSanitizer.sol

```diff
@@ -0,0 +1,328 @@
+    uint constant INCREMENT_CONTRACT_NONCE_GAS_COST_UPPER_BOUND = 200000;
+    // Code
+    uint constant ASSOCIATE_CODE_CONTRACT_GAS_COST_UPPER_BOUND = 200000;
+    uint constant REGISTER_CREATED_CONTRACT_GAS_COST_UPPER_BOUND = 200000;
+    uint constant GET_CODE_CONTRACT_ADDRESS_GAS_COST_UPPER_BOUND = 200000;
+    uint constant GET_CODE_CONTRACT_HASH_GAS_COST_UPPER_BOUND = 200000;
+    // Code copy retrieval, linear in code size
+    uint constant GET_CODE_CONTRACT_BYTECODE_GAS_COST_UPPER_BOUND = 200000;
+
+
+    /*
+     * Contract Variables
+     */
+
+    uint OVMRefund;
+
+    /*
+     * Modifiers
+     */
+
+
+    /*
+     * Constructor
+     */
+
+    /**
+     * @param _addressResolver Address of the AddressResolver contract.
+     */
+    constructor(
+        address _addressResolver
+    )
+        public
+        ContractResolver(_addressResolver)
+    {}
+
+    /*
+     * Gas Virtualization and Storage
+     */
+
+    // External Initialization and Retrieval Logic
+    function resetOVMRefund() external {
+        OVMRefund = 0;
+    }
+
+    function getOVMRefund() external view returns(uint) {
+        return OVMRefund;
+    }
+
+    // Internal Logic
+    function addToOVMRefund(uint _refund) internal {
+        OVMRefund += _refund;
+        return;
+    }
+
+    /** TODO UPDATE THIS DOCSTR
+     * Forwards a call to this proxy along to the actual state manager, and consumes any leftover gas up to _virtualGasCost.
+     */
+    function performSanitizedProxyAndRecordRefund(
+        uint _sanitizedGasCost,
+        uint _virtualGasCost
+    ) internal {
+        uint initialGas = gasleft();
+        address stateManager = resolveStateManager();
+
+        uint refund = _sanitizedGasCost - _virtualGasCost;
+        addToOVMRefund(refund);
+
+        bool success;
+        uint returnedSize;
+        uint returnDataStart;
+        assembly {
+            let initialFreeMemStart := mload(0x40)
+            let callSize := calldatasize()
+            mstore(0x40, add(initialFreeMemStart, callSize))
+            calldatacopy(
+                initialFreeMemStart,
+                0,
+                callSize
+            )
+            success := call(
+                gas(),
+                stateManager,
+                0,
+                initialFreeMemStart,
+                callSize,
+                0,
+                0
+            )
+            returnedSize := returndatasize()
+
+            // write the returndata to memory
+            returnDataStart := mload(0x40)
+            mstore(0x40, add(returnDataStart, returnedSize))
+            returndatacopy(
+                returnDataStart,
+                0,
+                returnedSize
+            )
+        }
+
+        // todo safemath negatives
+        GasConsumer gasConsumer = GasConsumer(resolveGasConsumer());
+        uint gasAlreadyConsumed = initialGas - gasleft();
+        uint gasLeftToConsume = _sanitizedGasCost - gasAlreadyConsumed;
```

Comment:
Just want to make sure it's not possible for this to underflow.

Reply:
We already have to make an assumption about the max gas an SM implementation will consume, which is why the constants above are called `*_GAS_COST_UPPER_BOUND`.  If that breaks, then we will have bigger issues than the underflow here, so I figured it's fine.